crates/red_knot_ide/src/lib.rs

```diff
@@ -1,3 +1,222 @@
+    /// The range that should be focused when navigating to the target.
+    ///
+    /// This is typically not the full range of the node. For example, it's the range of the class's name in a class definition.
+    ///
+    /// The `focus_range` must be fully covered by `full_range`.
+    focus_range: TextRange,
+
+    /// The range covering the entire target.
+    full_range: TextRange,
+}
+
+impl NavigationTarget {
+    pub fn file(&self) -> File {
+        self.file
+    }
+
+    pub fn focus_range(&self) -> TextRange {
+        self.focus_range
+    }
+
+    pub fn full_range(&self) -> TextRange {
+        self.full_range
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NavigationTargets(smallvec::SmallVec<[NavigationTarget; 1]>);
+
+impl NavigationTargets {
+    fn single(target: NavigationTarget) -> Self {
+        Self(smallvec::smallvec![target])
+    }
+
+    fn empty() -> Self {
+        Self(smallvec::SmallVec::new())
+    }
+
+    fn iter(&self) -> std::slice::Iter<'_, NavigationTarget> {
+        self.0.iter()
+    }
+
+    #[cfg(test)]
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl IntoIterator for NavigationTargets {
+    type Item = NavigationTarget;
+    type IntoIter = smallvec::IntoIter<[NavigationTarget; 1]>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a NavigationTargets {
+    type Item = &'a NavigationTarget;
+    type IntoIter = std::slice::Iter<'a, NavigationTarget>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl FromIterator<NavigationTarget> for NavigationTargets {
+    fn from_iter<T: IntoIterator<Item = NavigationTarget>>(iter: T) -> Self {
+        Self(iter.into_iter().collect())
+    }
+}
+
+pub trait HasNavigationTargets {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets;
+}
+
+impl HasNavigationTargets for Type<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        match self {
+            Type::FunctionLiteral(function) => function.navigation_targets(db),
+            Type::ModuleLiteral(module) => module.navigation_targets(db),
+            Type::Union(union) => union
+                .iter(db.upcast())
+                .flat_map(|target| target.navigation_targets(db))
+                .collect(),
+            Type::ClassLiteral(class) => class.navigation_targets(db),
+            Type::Instance(instance) => instance.navigation_targets(db),
+            Type::KnownInstance(instance) => instance.navigation_targets(db),
+            Type::StringLiteral(_)
+            | Type::AlwaysTruthy
+            | Type::AlwaysFalsy
+            | Type::IntLiteral(_)
+            | Type::BooleanLiteral(_)
+            | Type::LiteralString
+            | Type::BytesLiteral(_)
+            | Type::SliceLiteral(_) => self.to_meta_type(db.upcast()).navigation_targets(db),
+
+            Type::Dynamic(_)
+            | Type::SubclassOf(_)
```

Comment:
If this is a subclass-of a class type (not dynamic), then it should have the navigation targets of that class.

Reply:
you'll want to do something like

```rs
            Type::SubclassOf(subclass_of_type) => match subclass_of_type.subclass_of() {
                ClassBase::Class(class) => class.navigation_targets(db),
                ClassBase::Dynamic(_) => NavigationTargets::empty(),
            }
```